presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -529,92 +505,70 @@ private void addToHiveSplitSourceRoundRobin(List<Iterator<InternalHiveSplit>> it
             throws IOException
     {
         if (!pathMatchesPredicate(pathDomain, path)) {
-            return emptyIterator();
+            return Optional.empty();
         }
 
         boolean forceLocalScheduling = HiveSessionProperties.isForceLocalScheduling(session);
 
-        if (splittable) {
-            PeekingIterator<BlockLocation> blockLocationIterator = Iterators.peekingIterator(Arrays.stream(blockLocations).iterator());
-
-            return new AbstractIterator<InternalHiveSplit>()
-            {
-                private long chunkOffset = 0;
-
-                @Override
-                protected InternalHiveSplit computeNext()
-                {
-                    if (!blockLocationIterator.hasNext()) {
-                        return endOfData();
-                    }
-                    BlockLocation blockLocation = blockLocationIterator.peek();
+        ImmutableList.Builder<InternalHiveBlock> blockBuilder = ImmutableList.builder();
+        for (BlockLocation blockLocation : blockLocations) {
+            // clamp the block range
+            long blockStart = Math.max(start, blockLocation.getOffset());
+            long blockEnd = Math.min(start + length, blockLocation.getOffset() + blockLocation.getLength());
+            if (blockStart > blockEnd) {
+                // block is outside split range
+                continue;
+            }
+            if (blockStart == blockEnd && !(blockStart == start && blockEnd == start + length)) {
+                // skip zero-width block, except in the special circumstance: slice is empty, and the block covers the empty slice interval.
```

Comment:
Why do we need a special case for this?

Reply:
This seems to be a special case to generate a split for an empty file. Why not just check that up front?